lib/spack/spack/util/url.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""
+Utility functions for parsing, formatting, and manipulating URLs.
+"""
+
+import itertools as it
+
+from os.path import join as pathjoin, split as pathsplit
+
+from six import string_types
+from six.moves.urllib.parse import urlparse, ParseResult
+
+from spack.util.path import canonicalize_path
+
+
+class ParseResultWrapper(ParseResult):
+    def __setattr__(self, key, val):
+        object.__setattr__(self, key, val)
+
+    def __str__(self):
+        return ''.join((
+            super(ParseResultWrapper, self).__str__(),
+            ' ({})'.format(str(self.__dict__)) if self.__dict__ else ''))
+
+
+def _split_all(path):
+    result = []
+    a = path
+    old_a = None
+    while a != old_a:
+        (old_a, (a, b)) = a, pathsplit(a)
+
+        if a or b:
+            result.insert(0, b or '/')
+
+    return result
+
+
+def parse(url, scheme='file'):
+    """Parse a mirror url."""
```

Comment:
Better docstring -- document return type.  Also document why we need this instead of just urlparse.